### PR TITLE
Fix/hubspot company key

### PR DESCRIPTION
**Pull request by pauloouriques**
## Change description

This PR contains the code to fix the companyKey destination options issues when changing to another value.

### Development

- [x] Application changes have been tested appropriately

### Impact

- [x] Code follows company security practices and guidelines
- [x] Security impact of change has been considered
- [x] Performance impact of change has been considered
- [x] Possible migration needs considered (model migrations, config migrations, etc.)

Please explain any security, performance, migration, or other impacts if relevant:

### Code review

- [x] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached where applicable.
- [x] Relevant tags have been added to the PR (bug, enhancement, internal, etc.)

### CODE REVIEW

File: plugins/@grouparoo/hubspot/src/lib/export-contacts/exportRecord.ts [hunk 1]
```diff
@@ -180,6 +179,7 @@ async function syncAssociations(
   client: HubspotClient,
   contact: any,
   companyKey: string,
+  companyRecordField: string,
```
Comment (bleonard) [inline plugins/@grouparoo/hubspot/src/lib/export-contacts/exportRecord.ts:182]: but it's not actually using it?

**Review by bleonard — COMMENTED** [review]
I think I need some more information on this. I can't actually figure out how the new field is being used.
Ideally, we shouldn't have to set this. We'd only set the `companyKey` and that would trigger the behavior of adding a required field of `Company.Name` (for example) and not `Company.hs_object_id` by the way.
Then maybe `exportRecord` would use `Company.X` to know to make the mapping and not the option itself. That would also know it's changed between old and new, I suppose.

File: plugins/@grouparoo/hubspot/src/lib/export/options.ts [hunk 2]
```diff
@@ -65,34 +67,44 @@ class OptionsHandler {
...
+    }
+
+    const contactFields = await this.getObjectMatchNames("CONTACT");
+    out.companyRecordField.options = contactFields;
+    if (
+      !contactFields.includes(destinationOptions.companyRecordField?.toString())
+    ) {
+      destinationOptions.companyRecordField = null;
+    }
     return out;
```
Comment (edmundito) [inline plugins/@grouparoo/hubspot/src/lib/export/options.ts:108]: Maybe a more concise implementation of the function would be:

```ts
getContactOptions(...): Promise<DestinationOptionsMethodResponse> {
  const companyKey = searchableCompanyFields.includes(...) ? { ... } : null
  const contactFields = await ...
  const companyRecordField = contactFields.includes(...) ? { ... } : null
  
  return { companyKey, companyRecordField }
}
```

File: plugins/@grouparoo/hubspot/src/lib/export-contacts/destinationMappingOptions.ts [hunk 3]
```diff
@@ -108,3 +109,26 @@ export const getUserFields = async (
...
+  if (!options) {
+    return;
+  }
+  const companyKeys = ["companyKey", "companyRecordField"];
+  // needs either none or all keys
+  let count = 0;
+  for (const key of companyKeys) {
+    const value = options ? (options[key] || "").toString().trim() : "";
+    if (value.length > 0) {
+      options[key] = value;
```
Comment (edmundito) [inline plugins/@grouparoo/hubspot/src/lib/export-contacts/destinationMappingOptions.ts:123]: Is it necessary to change the value of the options here? Is this function about sanitizing or checking that the values are there?

File: plugins/@grouparoo/hubspot/src/lib/export-contacts/destinationMappingOptions.ts [hunk 4]
```diff
@@ -9,22 +9,23 @@ const importantFields = ["firstname", "lastname", "company"];
...
-          type = "phoneNumber";
+    const known = await getUserFields(client, appOptions);
+    if (destinationOptions?.companyRecordField) {
+      let index = -1;
+      for (let i = 0; i < known.length; i++) {
+        if (known[i].key === destinationOptions.companyRecordField) {
+          index = i;
           break;
+        }
+      }
```
Comment (edmundito) [inline plugins/@grouparoo/hubspot/src/lib/export-contacts/destinationMappingOptions.ts:23]:
```suggestion
      const index = known.findIndex(({ key }) => key === destinationOptions.companyRecordField);
```

File: plugins/@grouparoo/hubspot/src/lib/export-contacts/exportRecord.ts (hunk 1 above)
Comment (pauloouriques) [reply plugins/@grouparoo/hubspot/src/lib/export-contacts/exportRecord.ts:182]: actually, the way I did this time it's not using `Company.Name`, it's using like: `oldRecordProperties[companyRecordField]`.

File: plugins/@grouparoo/hubspot/src/lib/export-contacts/destinationMappingOptions.ts (hunk 3 above)
Comment (pauloouriques) [reply plugins/@grouparoo/hubspot/src/lib/export-contacts/destinationMappingOptions.ts:123]: I did both, just in case.

Comment (pauloouriques) [reply plugins/@grouparoo/hubspot/src/lib/export-contacts/destinationMappingOptions.ts:123]: I did both, just in case

**Pull request closed without merging**